public/dorequest.php

```diff
@@ -277,6 +345,104 @@ function DoRequestError(string $error, ?int $status = 200, ?string $code = null)
+            return DoRequestError('Access denied.', 403, 'access_denied');
+        }
+
+        $targetUser = User::firstWhere('User', $delegateTo);
+        if (!$targetUser) {
+            return DoRequestError("The target user couldn't be found.", 404, 'not_found');
+        }
+
+        $achievementIdsArray = explode(',', $achievementIdsInput);
+        $filteredAchievementIds = array_filter($achievementIdsArray, function ($id) {
+            return filter_var($id, FILTER_VALIDATE_INT) !== false;
+        });
+
+        // Fetch all achievements already awarded to the user.
+        $foundPlayerAchievements = PlayerAchievement::where('user_id', $targetUser->id)
+            ->with('achievement')
+            ->whereIn('achievement_id', $achievementIdsArray)
+            ->get();
+
+        $alreadyAwardedIds = [];
+
+        // Filter out achievements based on the hardcore flag and existing unlocks.
+        $filteredAchievementIds = array_filter($achievementIdsArray, function ($id) use (&$alreadyAwardedIds, $user, $foundPlayerAchievements, $hardcore) {
+            foreach ($foundPlayerAchievements as $foundPlayerAchievement) {
+                if ($foundPlayerAchievement->achievement_id == $id) {
+                    // If a record exists and it's already unlocked in hardcore mode, then
+                    // we won't attempt to re-award the achievement to the user.
+                    if ($hardcore && $foundPlayerAchievement->unlocked_hardcore_at !== null) {
+                        $alreadyAwardedIds[] = $foundPlayerAchievement->achievement_id;
+
+                        return false;
+                    }
+
+                    // If a record exists but it's a softcore unlock and hardcore mode is requested,
+                    // then we're going to add it to the list of attempted IDs to unlock.
+                    if ($hardcore && $foundPlayerAchievement->unlocked_hardcore_at === null) {
+                        return true;
+                    }
+
+                    // If a record exists (either hardcore or softcore) and softcore mode is requested,
+                    // then skip over this ID.
+                    if (!$hardcore) {
+                        $alreadyAwardedIds[] = $foundPlayerAchievement->achievement_id;
+
+                        return false;
+                    }
+
+                    if (!$foundPlayerAchievement->achievement->getCanDelegateUnlocks($user)) {
+                        return false;
+                    }
+                }
+            }
+
+            // If no PlayerAchievement record exists for this ID, it's always eligible for awarding.
+            return true;
+        });
+
+        $awardableAchievements = Achievement::whereIn('ID', $filteredAchievementIds)
+            ->with('game')
+            ->get()
+            ->filter(function ($achievement) use ($user) {
+                return $achievement->getCanDelegateUnlocks($user);
+            });
```

Comment:
`$achievement->getCanDelegateUnlocks($user)` is handled by the `filteredAchievementIds` logic above. This can be removed.

Reply:
Removed and refactored the filtering logic just a bit.